### PR TITLE
Add adc calibration support

### LINKED ISSUE #50: Perform calibration of ADC (Due: 2024-03-15)
The calibration will be performed with a python script that controls an SMU and the SPS. Separate calibrations will have to be performed for the voltage/current channel. Planned test setup for voltage

image

Outcome: Automated method of calibrating voltage/current channels using external SMU. Documentation should be written describing how to calibrate the device. A sane default calibration should be included in the firmware.

### PR DESCRIPTION
**Pull request by stevegtaylor**
---
name: Stephen Taylor
about: calibration eqn for the demo
title: 'Junior Specialist'
labels: ''
assignees: 'jmadden173'
reviewers: 'jmadden173'
---

**Name/Affiliation/Title**
Stephen Taylor, UCSC, staff

**Purpose of the PR**
Added a 'calibration' folder for calibration procedure. Integrated the positive voltage calibration for the purposes of the demo

**Development Environment**
Windows

**Test Procedure**
Run calibrate adc with a soil power sensor and the 2450 and see if it outputs reasonable values.

**Additional Context**
Closes #50

### CODE REVIEW

File: calibration/LICENSE
Comment (jmadden173) [inline calibration/LICENSE:1]: I don't think we need another license, so this could be removed

File: calibration/PingSMU.py
Comment (jmadden173) [inline calibration/PingSMU.py:1]: No docstrings

File: calibration/PingSPS.py
Comment (jmadden173) [inline calibration/PingSPS.py:1]: No file docstring and pdb is still included

File: calibration/data/config.yaml
Comment (jmadden173) [inline calibration/data/config.yaml:1]: Can you clarify if any of these settings are being used? They all look like they're from v1 of the board

File: calibration/least_squares_calibration.py
Comment (jmadden173) [inline calibration/least_squares_calibration.py:1]: Missing docstrings

File: calibration/linear_regression.py
Comment (jmadden173) [inline calibration/linear_regression.py:1]: Missing docstrings

File: calibration/least_squares_calibration.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+i_model = linear_model.LinearRegression()
+i_model.fit(data[i_input_cols], data["I_in"])
+
+eval_data = load_data(cfg, evalfiles)
+
+eval_data["V_fit"] = v_model.predict(eval_data[v_input_cols])
+eval_data["I_fit"] = i_model.predict(eval_data[i_input_cols])
+
+rld_v = load_rld("data/rl/20220826_200050_eval_v.rld", "V1", 0.1, 1, 0.05)
+rld_i = load_rld("data/rl/20220826_200050_eval_i.rld", "I1L", 45e-6, 454e-6, 22.7222e-6)
```
Comment (jmadden173) [inline calibration/least_squares_calibration.py:73]: The rdl files don't seem to match the range of the sps measurements.

File: calibration/pyMeasureQuickStart.py
Comment (jmadden173) [inline calibration/pyMeasureQuickStart.py:1]: No docstrings, what is this for?

File: calibration/recorder.py [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+        Initialises connection to serial port.
+
+        Parameters
+        ----------
+        port : str
+            Serial port of device
+        baudrate : int, optional
+            Baud rate for serial communication (default is 115200, STM32 functions at 115200)
+        xonxoff  : bool, optional
+            Flow control (default is on)
```
Comment (jmadden173) [inline calibration/recorder.py:58]: Shoot when I fixed the software flow control I forgot to update this doc

File: calibration/recorder.py
Comment (jmadden173) [inline calibration/recorder.py:1]: I don't think that there should be print statements within functions. It gets really annoying when trying to use a module/class and you get random print statements. I would add a verbose cli option that shows more detailed information if we are trying to debug something.

File: calibration/requirements.in
Comment (jmadden173) [inline calibration/requirements.in:1]: I don't think jupyterlab is required. Also I think we should convert the `requirements.in` to `requirements.txt` from my experience with trying this with `pip-compile`. If someone is on a different python version it completely breaks.

File: calibration/requirements.txt
Comment (jmadden173) [inline calibration/requirements.txt:1]: Remove

File: stm32/Src/ads.c [hunk 3]
```diff
@@ -107,6 +118,14 @@ double ADC_readVoltage(void){
   // char raw[45];
   // sprintf(raw, "Raw: %x %x %x Shifted: %f \r\n\r\n",rx_data[0], rx_data[1], rx_data[2], reading);
```
Comment (jmadden173) [inline stm32/Src/ads.c:119]: What are all these commented out lines for?

File: stm32/Src/ads.c
Comment (jmadden173) [inline stm32/Src/ads.c:1]: See general comment for more info

File: stm32/Src/examples/calibrate_adc.c [hunk 4]
```diff
@@ -122,37 +122,57 @@ int main(void)
...
+  
   while (1)
   {
 
     /* USER CODE END WHILE */    
 
     /* USER CODE BEGIN 3 */
-    HAL_UART_Receive(&huart1, (uint8_t *) controller_input, 2, 1000);
+    if (checked == false){ // this flag exists to make sure it only searches for the check command once
+      HAL_UART_Receive(&huart1, (uint8_t *) check_input, 5, 1000);
```
Comment (jmadden173) [inline stm32/Src/examples/calibrate_adc.c:145]: We shouldn't declare arbitrary timeout values. It should be a static global variable so that it can be changed for all UART functions. I don't understand why they are different between calls.

File: stm32/Src/examples/calibrate_adc.c [hunk 5]
```diff
@@ -122,37 +122,57 @@ int main(void)
...
-      controller_input[0] = 'n';
+      size_t measurement_size = ADC_measure(encoded_measurment); // Read the measurment, and store it's size in measurement_size (size int 64)
+
+      // send length
+      HAL_UART_Transmit(&huart1, (uint8_t *) &measurement_size, 1, 100);
+      // send data
+      HAL_UART_Transmit(&huart1, (uint8_t *) encoded_measurment, measurement_size, 1000);
+
+
+      //  // Define a buffer to hold the hex representation of each uint8_t value
```
Comment (jmadden173) [inline stm32/Src/examples/calibrate_adc.c:162]: Random comments?

File: stm32/Src/examples/example_adc.c [hunk 6]
```diff
@@ -143,9 +143,9 @@ int main(void)
     reading_len = sprintf(output, "Voltage: %f\r\n", voltage_reading);
     HAL_UART_Transmit(&huart1, (const uint8_t *) output, reading_len, HAL_MAX_DELAY);
 
-    current_reading = ADC_readCurrent();
-    reading_len = sprintf(output2, "Current: %f\r\n", current_reading);
-    HAL_UART_Transmit(&huart1, (const uint8_t *) output2, reading_len, HAL_MAX_DELAY);
+    // current_reading = ADC_readCurrent();
```
Comment (jmadden173) [inline stm32/Src/examples/example_adc.c:146]: Why are these commented out?

File: stm32/Src/examples/example_adc.c [hunk 7]
```diff
@@ -143,9 +143,9 @@ int main(void)
     reading_len = sprintf(output, "Voltage: %f\r\n", voltage_reading);
```
Comment (jmadden173) [inline stm32/Src/examples/example_adc.c:143]: I think we should be able to use `SysApp_Init` or whatever it's actually defined as to start the `UTIL_Timer` and give us access to `APP_LOG` instead of having to use `sprintf` which get messy.

File: stm32/Src/main.c [hunk 8]
```diff
@@ -39,6 +39,7 @@
 #include "phytos31.h"
 #include "rtc.h"
 #include "sensors.h"
+#include "transcoder.h"
```
Comment (jmadden173) [inline stm32/Src/main.c:42]: Why is this included in main?

**Review by jmadden173 — COMMENTED** [review]
What is the difference between `least_squares_calibration.py` and `linear_regression.py`?

I think there is a way to get a single calibration function for the entire range rather than have two separate calibration equations. The two equations will use more memory on the FRAM chip and more floating point computations. If the ADC outputs a 24bit full scale number there should be a way to get it to a uint32 with 0 as -3.3V and 2^24 as 3.3V.

File: calibration/LICENSE
Comment (stevegtaylor) [reply calibration/LICENSE:1]: removed

File: calibration/pyMeasureQuickStart.py
Comment (stevegtaylor) [reply calibration/pyMeasureQuickStart.py:1]: dunno, this is a file from the original repo. can be deleted, i dont use it

File: calibration/recorder.py (hunk 2 above)
Comment (stevegtaylor) [reply calibration/recorder.py:58]: should default be off?

**Comment by stevegtaylor** [comment]
least_squares_calibration.py is the Jupiter notebook file you made converted directly to a python file. linear regression.py is my file that i started from scratch

File: stm32/Src/examples/example_adc.c (hunk 6 above)
Comment (stevegtaylor) [reply stm32/Src/examples/example_adc.c:146]: i was testing things, added back

File: calibration/recorder.py
Comment (stevegtaylor) [reply calibration/recorder.py:1]: do you mean in the socket/serial connect functions? those are intentional to make debugging connection issues easier.

File: calibration/least_squares_calibration.py (hunk 1 above)
Comment (stevegtaylor) [reply calibration/least_squares_calibration.py:73]: least_squares_calibration isn't used in my calibration

File: calibration/data/config.yaml
Comment (stevegtaylor) [reply calibration/data/config.yaml:1]: haven't touched this file, so they probably are from v1 of the board

Comment (jmadden173) [inline calibration/data/config.yaml:1]: Is this file needed?

File: git
Comment (jmadden173) [inline git:1]: Remove

File: python
Comment (jmadden173) [inline python:1]: Remove

File: stm32/cpplint.cfg
Comment (jmadden173) [inline stm32/cpplint.cfg:1]: File name should be all uppercase or else cpplint will not work

File: calibration/README.MD
Comment (jmadden173) [inline calibration/README.MD:1]: This file should be flushed out and explain what each of the files are and how to use them

File: calibration/model_testing.py
Comment (jmadden173) [inline calibration/model_testing.py:1]: This file looks like its not needed anymore

File: calibration/requirements.txt
Comment (jmadden173) [inline calibration/requirements.txt:1]: I don't think `scipy` and `numpy` are included in this requirements.txt

**Review by jmadden173 — CHANGES_REQUESTED** [review]
Run linter on python code following the `soil_power_sensor_protobuf` module

As talked about move calibration traces to separate repos.

I think we should remove calibration python files that aren't use so make it more accessible. Possible archive them in Google Drive so that if we have a use case in the future they are there.

**Comment by jlin143** [comment]
Here was the fix:

1. HAL_Delay(60) // at least 50.01 ms after configure in readvoltage readcurrent
2. re-enable calibration and make sure that the readcurrent calibration is not commented out --> change to the const not the defines
3. single shot mode: adjust the argument passed to ADC configure to not have the second least bit high ( & ~0b10)
4. consider adjusting the Data Rate (DR field in the config register) if the sample period (20 SPS ~ 50.01 ms) is too slow.
